### PR TITLE
[rfr] More User and Node

### PR DESCRIPTION
**Pull request by sloria**
Ports a lot more of the User and Node models, including

- Contributor related methods, e.g. `add_contributor`, add_contributors`, `set_visible`
- Permission related methods, e.g. `add_permission`, `has_permissions`
- Account registration and confirmation methods

And a number of fixes:

- A node's creator is automatically added as an admin contributor
- Fix get_permissions and add_log

### CODE REVIEW

**Comment by sloria** [comment]
@cwisecarver This is ready for review.

File: osf_models/models/contributor.py [hunk 1]
```diff
@@ -22,3 +22,11 @@ class InstitutionalContributor(AbstractBaseContributor):
...
     class Meta:
         unique_together = ('user', 'institution')
+
+class RecentlyAddedContributor(models.Model):
+    user = models.ForeignKey('OSFUser')  # the user who added the contributor
+    contributor = models.ForeignKey('OSFUser', related_name='recently_added_by')  # the added contributor
+    date_added = models.DateTimeField(auto_now=True)
+
+    class Meta:
+        unique_together = ('user', 'contributor')
```
Comment (cwisecarver) [inline osf_models/models/contributor.py]: Without a reference to the node, wouldn't this disallow `me` adding `you` to multiple of my projects?

File: osf_models/models/mixins.py [hunk 2]
```diff
@@ -71,7 +71,7 @@ def add_log(self, action, params, auth, foreign_user=None, log_date=None, save=T
         if self.logs.count() == 1:
             self.date_modified = log.date.replace(tzinfo=pytz.utc)
         else:
-            self.date_modified = self.logs[-1].date.replace(tzinfo=pytz.utc)
+            self.date_modified = self.logs.order_by('-date')[0].date.replace(tzinfo=pytz.utc)
```
Comment (cwisecarver) [inline osf_models/models/mixins.py]: Django gets very upset if you deal with naive datetimes. In the migrations, I'm coercing everything to UTC prior to inserting it. Or at least, that is my intention.

There's also a `.first()` method on querysets. You could change the [default sort](https://docs.djangoproject.com/ja/1.9/ref/models/options/#ordering) for Nodelogs and then do `self.logs.first().date`

File: osf_models/models/node.py [hunk 3]
```diff
@@ -183,20 +206,35 @@ def get_absolute_url(self):
...
         contrib = user.contributor_set.get(node=self)
         perm = []
-        if contrib.admin:
-            perm.append('admin')
-        if contrib.write:
-            perm.append('write')
         if contrib.read:
-            perm.append('read')
-        return []
+            perm.append(READ)
```
Comment (cwisecarver) [inline osf_models/models/node.py]: 👍  much better than magic strings

File: osf_models/models/node.py [hunk 4]
```diff
@@ -239,16 +277,174 @@ def add_tag_log(self, tag, auth):
...
+        :param bool save: Save after adding contributor
+        :returns: Whether contributor was added
+        """
+        MAX_RECENT_LENGTH = 15
+
+        # If user is merged into another account, use master account
+        contrib_to_add = contributor.merged_by if contributor.is_merged else contributor
+        if not self.is_contributor(contrib_to_add):
+
+            contributor_obj, created = Contributor.objects.get_or_create(user=contrib_to_add, node=self)
```
Comment (cwisecarver) [inline osf_models/models/node.py]: Minor: You can use the relative relationship manager instead of importing.

`contributor_obj, created = self.contributors.get_or_create(user=contrib_to_add, node=self)`

File: osf_models/models/node.py [hunk 5]
```diff
@@ -239,16 +277,174 @@ def add_tag_log(self, tag, auth):
...
+            permissions = permissions or DEFAULT_CONTRIBUTOR_PERMISSIONS
+            for perm in permissions:
+                setattr(contributor_obj, perm, True)
+            contributor_obj.save()
+
+            # Add contributor to recently added list for user
+            if auth is not None:
+                RecentlyAddedContributor = apps.get_model('osf_models.RecentlyAddedContributor')
+                user = auth.user
+                recently_added_contributor_obj, created = RecentlyAddedContributor.objects.get_or_create(
```
Comment (cwisecarver) [inline osf_models/models/node.py]: Same here with the relationship manager.

File: osf_models/models/contributor.py (hunk 1 above)
Comment (sloria) [inline osf_models/models/contributor.py]: This is the through table for the `recently_added` field of `User`. This a relationship between users and doesn't have to do with nodes. A User "A" can only have one entry for a User "B" in their recently_added list.

File: osf_models/models/node.py [hunk 6]
```diff
@@ -239,16 +277,174 @@ def add_tag_log(self, tag, auth):
...
+                    contributor=contrib_to_add
+                )
+                recently_added_contributor_obj.date_added = dt.datetime.utcnow()
+                recently_added_contributor_obj.save()
+                count = user.recently_added.count()
+                if count > MAX_RECENT_LENGTH:
+                    difference = count - MAX_RECENT_LENGTH
+                    for each in user.recentlyaddedcontributor_set.order_by('date_added')[:difference]:
+                        each.delete()
+                    user.save()
```
Comment (cwisecarver) [inline osf_models/models/node.py]: I don't think you have to save here. I'm pretty sure `.delete()` will run a query every time you call it. Though, I believe the ORM will stop you actually running SQL if nothing changed.

[Also, you can call `.delete()` on a queryset to kill them all in one query](http://stackoverflow.com/questions/1851197/django-delete-all-but-last-five-of-queryset#answer-20996456).

File: osf_models/models/user.py [hunk 7]
```diff
@@ -206,15 +209,24 @@ class OSFUser(DirtyFieldsMixin, GuidMixin, BaseModel, AbstractBaseUser, Permissi
...
     # recently_added = fields.ForeignField("user", list=True)
-    recently_added = models.ManyToManyField('self', symmetrical=False)
+    recently_added = models.ManyToManyField('self',
+                                            through=RecentlyAddedContributor,
+                                            through_fields=('user', 'contributor'),
+                                            symmetrical=False)
+
+    def get_recently_added(self):
+        return (
+            each.contributor
```
Comment (cwisecarver) [inline osf_models/models/user.py]: I believe you could do `return tuple(self.recentlyaddedcontributor_set.values('contributor', flat=True))`

File: osf_models/models/node.py (hunk 4 above)
Comment (sloria) [inline osf_models/models/node.py]: That doesn't quite work because `contributors` is a m2m with `User`. In this, case I want to create a `Contributor` object directly so that I can set visibility on `permissions` on it.

Comment (cwisecarver) [inline osf_models/models/node.py]: Ah, I forgot that didn't work when specifying a through table.

File: osf_models/models/node.py (hunk 5 above)
Comment (cwisecarver) [inline osf_models/models/node.py]: pls ignore

File: osf_models/models/node.py [hunk 8]
```diff
@@ -239,16 +277,174 @@ def add_tag_log(self, tag, auth):
...
+
+            # Add contributor to recently added list for user
+            if auth is not None:
+                RecentlyAddedContributor = apps.get_model('osf_models.RecentlyAddedContributor')
+                user = auth.user
+                recently_added_contributor_obj, created = RecentlyAddedContributor.objects.get_or_create(
+                    user=user,
+                    contributor=contrib_to_add
+                )
+                recently_added_contributor_obj.date_added = dt.datetime.utcnow()
```
Comment (cwisecarver) [inline osf_models/models/node.py]: Do you want to change the `date_added` if the recently_added_contributor already existed?

Comment (sloria) [inline osf_models/models/node.py]: Yes, it should always be updated. That way, when the frontend displays recently-added contribs, the contribs are displayed in the correct order.

File: osf_models/models/user.py (hunk 7 above)
Comment (sloria) [inline osf_models/models/user.py]: Tried that, but it didn't work. `values('contributor')` returns a list of dictionaries (`{contributor': 123}') and `values_list('contributor', flat=True)` returns a list of integer PKs.

File: osf_models/models/node.py (hunk 6 above)
Comment (sloria) [inline osf_models/models/node.py]: Good call on removing the `save()`. I can't call `.delete()` on this queryset because I'm using a `limit()` here, which leads to this error:

```
|| osf_models_tests/test_user.py:766: in test_recently_added_length
||     auth=auth
|| ../osf_models/osf_models/models/node.py:353: in add_contributor
||     user.recentlyaddedcontributor_set.order_by('date_added')[:difference].delete()
|| ../../../miniconda/envs/osf/lib/python2.7/site-packages/django/db/models/query.py:581: in delete
||     "Cannot use 'limit' or 'offset' with delete."
|| E   AssertionError: Cannot use 'limit' or 'offset' with delete
```

**Comment by sloria** [comment]
@cwisecarver I've rebased on top of master and addressed your comments.

**Comment by cwisecarver** [comment]
Thanks. Sorry for all the red herrings. I guess I don't know the ORM as well as I thought I did. We're getting into a lot of edge-casey areas though. lgtm to merge whenever you feel like it.

**Comment by sloria** [comment]
No worries. Getting the tests passing. I'll merge when that's done.
